src/org/exist/xquery/functions/fn/FunDocAvailable.java

```diff
@@ -35,79 +33,83 @@
+                    new SequenceType[]{
+                            new FunctionParameterSequenceType("document-uri", Type.STRING,
+                                    Cardinality.ZERO_OR_ONE, "The document URI")
+                    },
+                    new FunctionReturnSequenceType(Type.BOOLEAN, Cardinality.EXACTLY_ONE,
+                            "true() if the document is available, false() otherwise"));
 
-    public FunDocAvailable(XQueryContext context) {
+    public FunDocAvailable(final XQueryContext context) {
         super(context, signature);
     }
 
-    /**
-     * @see org.exist.xquery.Function#getDependencies()
-     */
+    @Override
     public int getDependencies() {
         return Dependency.CONTEXT_SET;
     }
 
-    /**
-     * @see org.exist.xquery.Expression#eval(Sequence, Item)
-     */
-    public Sequence eval(Sequence contextSequence, Item contextItem)
+    @Override
+    public Sequence eval(final Sequence contextSequence, final Item contextItem)
             throws XPathException {
         if (context.getProfiler().isEnabled()) {
             context.getProfiler().start(this);
             context.getProfiler().message(this, Profiler.DEPENDENCIES,
-                "DEPENDENCIES", Dependency.getDependenciesName(this.getDependencies()));
-            if (contextSequence != null)
-                {context.getProfiler().message(this, Profiler.START_SEQUENCES,
-                    "CONTEXT SEQUENCE", contextSequence);}
-            if (contextItem != null)
-                {context.getProfiler().message(this, Profiler.START_SEQUENCES,
-                    "CONTEXT ITEM", contextItem.toSequence());}
+                    "DEPENDENCIES", Dependency.getDependenciesName(this.getDependencies()));
+            if (contextSequence != null) {
+                context.getProfiler().message(this, Profiler.START_SEQUENCES,
+                        "CONTEXT SEQUENCE", contextSequence);
+            }
+            if (contextItem != null) {
+                context.getProfiler().message(this, Profiler.START_SEQUENCES,
+                        "CONTEXT ITEM", contextItem.toSequence());
+            }
         }
-        Sequence result;
+
+        Sequence result = BooleanValue.FALSE;
         final Sequence arg = getArgument(0).eval(contextSequence, contextItem);
-        if (arg.isEmpty()) {
-            result = BooleanValue.FALSE;
-        } else {
+        if (!arg.isEmpty()) {
             final String path = arg.itemAt(0).getStringValue();
             try {
                 result = BooleanValue.valueOf(DocUtils.isDocumentAvailable(this.context, path));
             } catch (final XPathException e) {
-                e.prependMessage(ErrorCodes.FODC0005, "");
-                throw e;
+                if (e.getCause() instanceof SAXParseException) {
+                    result = BooleanValue.FALSE;
+                } else if(e.getMessage().contains("is a binary resource, not an XML document")) {
```

Comment:
I completely agree. However there is no better fix without either changing the API (which we can't do for a hotfix release) or duplicating code. I think a better fix could come in future. This was just designed for 3.1.1 and it improves the current situation.